Fix mapping of intermediate inherited classes

ByronPogson opened the pull request:
If two mapped classes A and B are in a class hierarchy where B indirectly inherits from A and mappings are declared for A and B, but not for the classes between A and B, NHibernate doesn't find any members declared in the classes between A and B that are mapped in B.

Fixes #1358

gliljas commented on src/NHibernate/Mapping/ByCode/IModelInspector.cs:
```diff
@@ -7,7 +7,7 @@ public interface IModelInspector
 		bool IsRootEntity(System.Type type);
 		bool IsComponent(System.Type type);
-		bool IsEntity(System.Type type);
+		bool IsEntity(System.Type type, bool explicitlyDeclared = false);
```
Since the interface is changed, an IsDeclaredEntity method would make more sense, or is there something I'm not seeing?

ByronPogson replied:
Given that default parameters aren't currently used in the code base this makes sense to expose two different methods, will update

gliljas commented on src/NHibernate.Test/NHSpecificTest/NH3992/FixtureByCode.cs:
```diff
@@ -0,0 +1,71 @@
+		}
+
+		[Test]
+		public void Test_InhertiedClasses_AllMapsAtTopLevel()
```
The test should test the generated mapping, not the outcome of the mapping.

ByronPogson replied:
Seems obvious now that someone points it out! Will update and re-push

fredericDelaporte commented:
Thanks for your contribution,

May you please reword your commits for including the issue number?

Something like:

- NH-3992 - Intermediate inherited classes test cases
- NH-3992 - Intermediate inherited classes fix

This allows the commits to be associated in Jira, as explained in [contributing guidelines](https://github.com/nhibernate/nhibernate-core/blob/master/CONTRIBUTING.md).

This PR introduces default values for parameters. It does not look to me as an accepted practice in NHibernate, since I can not find any of them (excepted in test project, where I have put some recently myself). And well, following @gliljas recommendation would avoid it.

ByronPogson commented:
Thanks for the feedback guys, I've made the changes are requested

fredericDelaporte commented:
Thanks,

I wonder whether `IsMappedClass` would be a more explicit name than `IsDeclaredEntity` or not. Note that I am not used to mapping by code, I personally only use hbm currently. That is how I understand that new method from my hbm mapping perspective.

ByronPogson commented:
The original existing method that was being called was IsEntity, so I think IsDeclaredEntity was suggested because it fits with that . Happy to change if others agree, I was just going with the suggestion from @gliljas

hazzik commented on src/NHibernate/Mapping/ByCode/ModelMapper.cs:
```diff
@@ -736,7 +736,7 @@ private System.Type GetEntityBaseType(System.Type type)
+				if (modelInspector.IsDeclaredEntity(analyzingType))
 				{
 					return analyzingType;
 				}
```
Next line shall be also changed to IsDeclaredEntity

hazzik replied:
(with tests)

ByronPogson replied:
Of course :) Just trying to work out a test that will make sure that line is hit

hazzik replied:
I think it would be triggered with a subclass of a subclass 

```
GrandSubclass -> Subclass -> Unmapped -> RootEntity
```

hazzik replied:
You need to have an interface as your `BaseEntity`

hazzik commented on src/NHibernate/Mapping/ByCode/SimpleModelInspector.cs:
```diff
@@ -31,7 +31,13 @@ public override bool IsRootEntity(System.Type entityType)
+
+			public bool IsDeclaredEntity(System.Type type)
+			{
+				return RootEntities.Contains(type) || HasDelayedEntityRegistration(type);
```
I doubt that HasDelayedEntityRegistration is required here. Could you please check with test that this is required?

If this is not required, then this method is equivalent to `IsRootEntity`

hazzik replied:
I think it would be required with a subclass of a subclass 

```
GrandSubclass -> Subclass -> Unmapped -> RootEntity
```

hazzik commented:
I've replaced `IsDeclaredEntity` with `IsRootEntity` and your tests still work

hazzik commented on src/NHibernate.Test/NHSpecificTest/NH3992/FixtureByCode.cs:
```diff
@@ -0,0 +1,92 @@
+	/// <summary>
+	/// Fixture using 'by code' mappings
+	/// </summary>
+	public class ByCodeFixture : TestCaseMappingByCode
```
Do not need to inherit TestCaseMappingByCode here.

hazzik commented on src/NHibernate.Test/NHSpecificTest/NH3992/FixtureByCode.cs:
```diff
@@ -0,0 +1,92 @@
+	/// </summary>
+	public class ByCodeFixture : TestCaseMappingByCode
+	{
+		protected override HbmMapping GetMappings()
```
Not needed as per comment above

ByronPogson commented:
Sorry for the delay. I have now made the changes above in regards to a) unit testing around interface inheritor and b) using the existing IsRootEntity method instead

ByronPogson commented:
@hazzik I think this is occurring because the test is setting up it's classes incorrectly. The Test has the following classes:
abstract Animal
abstract Reptile: Animal
Lizard: Reptile
abstract Mammal: Animal
Dog: Mammal
Cat: Mammal

These are then _all_ passed to the ConventionModelMapper.CompileMappingsFor() which throws the following exception:
System.NotSupportedException : The class NHibernate.Test.NHSpecificTest.NH2691.Reptile can't be instatiated and does not have mapped subclasses; 
possible solutions:
- don't map the abstract class
- map its subclasses.

This was probably fine when the intermediate classes were ignored but now they're not this is causing problems with this specific test. I can't find any documentation around how ConventionModelMapper should be used and what the intended usage is. All other tests using this are still passing. Removing the abstract intermediates (as suggested by the error) resolves the issue for this test. Is that sufficient or is there some documentation around which classes should be passed to this mapper?

hazzik commented:
I think it's a valid use case. The code generated is like following:

<?xml version="1.0" encoding="utf-8"?>
<hibernate-mapping xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" namespace="NHibernate.Test.NHSpecificTest.NH2691" assembly="NHibernate.Test, Version=5.0.0.0, Culture=neutral, PublicKeyToken=null" xmlns="urn:nhibernate-mapping-2.2">
  <class name="Animal" abstract="true">
    <id name="Id" type="Int32" />
    <discriminator />
    <property name="Description" />
    <property name="Sequence" />
  </class>
  <subclass name="Reptile" extends="Animal">
    <property name="BodyTemperature" />
  </subclass>
  <subclass name="Mammal" extends="Animal">
    <property name="Pregnant" />
    <property name="BirthDate" />
  </subclass>
  <subclass name="Lizard" extends="Animal">
    <property name="BodyTemperature" />
  </subclass>
  <subclass name="Dog" extends="Animal">
    <property name="Pregnant" />
    <property name="BirthDate" />
  </subclass>
  <subclass name="Cat" extends="Animal">
    <property name="Pregnant" />
    <property name="BirthDate" />
  </subclass>
</hibernate-mapping>

hazzik commented:
> I doubt that HasDelayedEntityRegistration is required here. Could you please check with test that this is required?
> If this is not required, then this method is equivalent to IsRootEntity
> I think it would be required with a subclass of a subclass
> 
> ```
> GrandSubclass -> Subclass -> Unmapped -> RootEntity
> ```
It seems that the tests shows that this is required

ByronPogson commented:
I've added in a further test to ensure that the mapping where intermediate classes are abstract and reverted to the previous method including the HasDelayedEntityRegistration

hazzik commented:
I saw that you pushed this commit, so my last to comments affect it. I mean
that after this commit it probably still possible to get incorrect mappings
built
On Tue, 29 Aug 2017 at 6:44 PM, ByronPogson <notifications@github.com>
wrote:

> I've added in a further test to ensure that the mapping where intermediate
> classes are abstract and reverted to the previous method including the
> HasDelayedEntityRegistration
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nhibernate/nhibernate-core/pull/609#issuecomment-325570484>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAI1l4Pjee5zG1pA1B0jntLJzHpQYMZyks5sc7NpgaJpZM4NJtL6>
> .
>

ByronPogson commented:
The comment that HasDelayedEntityRegistration is required? I've put that check back in. What other circumstances need to be tested?

hazzik commented:
Oh. My two comments was not send through. :(

hazzik commented:
I think it will still fail in some longer inheritance chains:

Mapped -> Unmapped -> Mapped -> Root
Mapped -> Mapped -> Unmapped -> Root (this most likely works)
Mapped -> Unmapped -> Mapped -> Unmapped -> Root

For these longer chains, could you please add tests to ensure that children "extend" correct parents?

ByronPogson commented:
Those chains you suggested don't work in the current code when calling the ConventionModelMapper with only the mapped classes. The unmapped classes don't get set as the "extend" and their properties aren't mapped. Looking through the code this is because the ConventionModelMapper uses the SimpleModelInspector and the IsDeclaredEntity method that is implemented (SimpleModelInspector:706) is identical to the IsEntity method. I suspect that either this method should return the declaredRelult parameter immediately or a isDeclaredEntity func should be added on line 174 with the default being set to return declared which could then be overridden as required.

This along however doesn't solve the issue for the ConventionModelMapper as presently no classes are registered as explicitly declared so none of them get returned as declared when the IModelExplicitDeclarationsHolder is declared model is examined. One solution I have is to append events in the ConventionModelMapper for each of the after class mapping events to add these into the declarations holder but this feels wrong as it's only addressing this issue and doesn't result in anything else being populated which seems outside the scope of this particular issue. Can anyone suggest a better solution to this?

ByronPogson commented:
We're keen to get this included in 5.1, can anyone provide feedback on the above so we can finish this one off?

ByronPogson commented:
@hazzik can you please provide guidance on this?

@billpoole-mi will be taking this one over

hazzik commented:
>It took me some time to figure it out.

It took me almost two years to figure out what the fix should look like 😅.

hazzik commented:
Squashed commits to reduce noise. And force pushed. No changes to last review by @fredericDelaporte
